### ISSUE TITLE
Send Presentation Error

### ISSUE THREAD

**Issue by yunxi-zhang** [hyperledger/aries-cloudagent-python#1598]
Hello, not sure if I should raise this ticket here or in Ursa repo.

When calling the API: present-proof/records/{{supplier_presentation_exchange_id}}/send-presentation, the body payload works for the 1st time but fails in the second time. Below shows the body payload that failed.

{
```
    "requested_attributes": {
        "attrProp1": {
            "cred_id": "57bfb3bd-dbf3-48da-aaf3-8d5891a4f8ab",
            "revealed": true
        },
        "attrProp2": {
            "cred_id": "57bfb3bd-dbf3-48da-aaf3-8d5891a4f8ab",
            "revealed": true
        }
    },
    "requested_predicates": {
        "predProp1": {
            "cred_id": "57bfb3bd-dbf3-48da-aaf3-8d5891a4f8ab",
            "timestamp": 1642073893
        }
    },
    "self_attested_attributes": {},
    "trace": false
}
```
This is the error message I've got from the API call, which is not clear enough to explain what's wrong. I see this error message comes from the Ursa repo.
`400: Error when constructing proof: Error: Invalid structure. Caused by: UrsaCryptoError: Predicate is not satisfied. CommonInvalidStructure.`

Can anyone help explain what it means?

**Comment by swcurran** [comment]
It's odd that it is failing with an Ursa message.  That implies then when the presentation is sent the first time, the part of the protocol state object that is the presentation is updated somehow such that it fails in Ursa.  Not sure why that would happen.  Is there any way for you to look at the data passed into the Ursa call on each call and see the difference.

That said, I'm not sure sending the same presentation twice should work.  If the presentation has been sent, does it make sense to send it again?

**Comment by yunxi-zhang** [comment]
@swcurran, I didn't explain well on what I've done. Actually, I've created two proof requests. **Proof Request 1** with the same payload structure as shown in my ticket did go through this API call successfully. The error I've got was after I created **Proof Request 2** sent from the verifier, and when the holder tried to send the proof presentation, then I got this error, which doesn't make sense to me, because the process was exactly the same and handled by my frontend app. I'll clean my environment again and reset up the whole local environment and test it again.

**Comment by swcurran** [comment]
Ah...got it.  That is not good -- that definitely should not happen.  If you can provide an example, please do.  This is easily tested in the Alice/Faber demo.  Just tried it and ran about 10 presentations in a row and all is worked.

**Comment by yunxi-zhang** [comment]
@swcurran, I reset up the whole environment and test it again, and the same issue still occurs after I created 2 proof requests.  In the proof request, I have 2 attribute request and 1 predicate request. The only differences between them are (1) the predicate's value is a bit different (e.g. 1st predicate said DOB < 2000 and 2nd predicate said DOB < 1990) and (2) the time stamp used, and the time stamp is created when a proof is requested which is a simple logic implemented in my frontend. I printed out the body payload used for the `send-presentation` API for the 2 tries as shown below.

First body payload was successful, I forgot to take a screenshot of the successful response.
![Screenshot 2022-01-16 at 20 56 49](https://user-images.githubusercontent.com/16185797/149677969-3e68a706-5be9-4536-8226-d5c9e199b992.png)

Second body payload caused errors.
![Screenshot 2022-01-16 at 20 58 17](https://user-images.githubusercontent.com/16185797/149678003-2b578eb0-889a-430b-8445-47fde72aa494.png)

Could you ask someone to test this case and see if this issue can be reproduced? Please see the ACA-Py version I use in my ticket.

**Comment by swcurran** [comment]
To reproduce this, could you please share the JSON of the credential that you shared, the data in the credential (although that hopefully doesn't matter) and the two proof requests.  Then we can replicate the same scenario using Alice/Faber.

Have you tried running only the second proof request/proof?  Perhaps it is something in that proof request.

**Comment by yunxi-zhang** [comment]
@swcurran , see below JSON information.

1. Credential JSON:
```
{
    "results": [
        {
            "referent": "91f4aa0f-b456-463c-ad71-0ebc7c4576d1",
            "attrs": {
                "expire_date": "2030",
                "issue_date": "2020",
                "last_name": "LName1",
                "date_of_birth": "1990",
                "first_name": "FName1"
            },
            "schema_id": "MtSBSkitb28PSoCj9EpSDs:2:passport:0.16",
            "cred_def_id": "MtSBSkitb28PSoCj9EpSDs:3:CL:14069:passport",
            "rev_reg_id": "MtSBSkitb28PSoCj9EpSDs:4:MtSBSkitb28PSoCj9EpSDs:3:CL:14069:passport:CL_ACCUM:364824ea-c227-480e-8a35-ca803c066d78",
            "cred_rev_id": "1"
        }
    ]
}
```

2. Proof presentation payload 1 (payload 2 only uses a different time stamp)

```
{
    "requested_attributes": {
        "attrProp1": {
            "cred_id": "91f4aa0f-b456-463c-ad71-0ebc7c4576d1",
            "revealed": true
        },
        "attrProp2": {
            "cred_id": "91f4aa0f-b456-463c-ad71-0ebc7c4576d1",
            "revealed": true
        }
    },
    "requested_predicates": {
        "predProp1": {
            "cred_id": "91f4aa0f-b456-463c-ad71-0ebc7c4576d1",
            "timestamp": 1642300578
        }
    },
    "self_attested_attributes": {},
    "trace": false
}
```

I've tried proof presentations for twice. 1st worked, but 2nd failed.

**Comment by yunxi-zhang** [comment]
@swcurran, any update?

**Comment by swcurran** [comment]
Sorry - no. I've not had time to try this.   It's a very weird one.

**Comment by swcurran** [comment]
Going through old issues and saw this one again.  I still am not sure what could be going on, but did want to confirm this. I assume that you are generating and sending a new presentation based off the second presentation request -- correct?  E.g. you aren't literally sending the same presentation on the second run?  That would generate an error -- as expected. The presentation request sent will have a unique nonce per request, and so using the same presentation twice will generate an error.

Thanks.

@ianco -- could you please run through a scenario like this?

**Comment by yunxi-zhang** [comment]
@swcurran, yes, the new presentation was based off the second presentation request.

**Comment by swcurran** [comment]
I tested this the easiest way I know how and it works fine:

- Start von-network instance
- Start indy-tails-service instance
- Run Faber: `./run_demo faber --revocation`
- Run Alice: `./run_demo alice`
- Connect the two -- copy Faber invite and paste at prompt from Alice
- Faber: Menu item 1 -- issue credential -- works
- Faber: Menu item 2 -- request proof -- works
- Faber: Menu item 2 -- request proof -- works
- Faber: Menu item 5 - revoke credential -- works
- Faber: Menu item 2 -- request proof -- works -- verification fails
- Faber: Menu item 1 -- issue credential -- works
- Faber: Menu item 2 -- request proof -- works
- Faber: Menu item 2 -- request proof -- works

That is a single program requesting the same proof content, but different proof requests multiple times. I also did a try without revocation and it worked fine as well.

I can't imagine this doesn't work -- it's something that has been working for many years.  My guess is that there is something in the setup you have or the sequence of steps.

I guess a full log of the steps you are doing might help.

**Comment by yunxi-zhang** [comment]
@swcurran, thanks for the update. Strangely, I can't reproduce this issue either in my environment at the moment. I'll close this ticket for now. If got new findings, would reopen it.
